# AB Test and Bidding Methods Conversion Comparison

## Business Problem
### Facebook recently introduced a new bidding type, 'average bidding', as an alternative to the existing bidding type called 'maximum bidding'.
### One of our customers, bombabomba.com, decided to test this new feature and would like to do an A/B test to see if average bidding converts more than maximum bidding.
### The A/B test has been going on for 1 month and bombabomba.com is now waiting for you to analyze the results of this A/B test. The ultimate success criterion for Bombabomba.com is Purchase. Therefore, the focus should be on the Purchase metric for statistical testing.

## Dataset Story

### In this data set, which includes the website information of a company, there is information such as the number of advertisements that users see and click, as well as earnings information from here. There are two separate data sets, the control and test groups. These datasets are in separate sheets of the ab_testing.xlsx excel. Maximum Bidding for the control group, Average for the test group Bidding has been applied.

* Impression: Ad views
* Click: Number of clicks on the displayed ad
* Purchase: The number of products purchased after the ads clicked
* Earning: Earnings after purchased products

In [1]:
# Task 1: Preparing and Analyzing Data

In [2]:
# Step 1: ab_testing_data.xlsx adlı kontrol ve test grubu verilerinden oluşan veri setini okutunuz. Kontrol ve test grubu verilerini ayrı değişkenlere atayınız.

In [39]:
import itertools
import matplotlib.pyplot as plt
import pandas as pd
import math
import numpy as np
import statsmodels.stats.api as sms
import scipy.stats as st
from scipy.stats import shapiro, levene, ttest_ind, mannwhitneyu
from statsmodels.stats.proportion import proportions_ztest
from sklearn.preprocessing import MinMaxScaler
import warnings
warnings.filterwarnings("ignore") 

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', 500)
pd.set_option('display.expand_frame_repr', False)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

In [11]:
df_control = pd.read_excel("ab_testing.xlsx", sheet_name="Control Group")
df_test = pd.read_excel("ab_testing.xlsx", sheet_name="Test Group")

In [12]:
df_control.head()

Impression      Click  Purchase    Earning  Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13
0  82529.45927 6090.07732 665.21125 2311.27714         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
1  98050.45193 3382.86179 315.08489 1742.80686         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
2  82696.02355 4167.96575 458.08374 1797.82745         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
3 109914.40040 4910.88224 487.09077 1696.22918         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
4 108457.76263 5987.65581 441.03405 1543.72018         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN

In [13]:
df_test.head()

Impression      Click  Purchase    Earning  Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13
0 120103.50380 3216.54796 702.16035 1939.61124         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
1 134775.94336 3635.08242 834.05429 2929.40582         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
2 107806.62079 3057.14356 422.93426 2526.24488         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
3 116445.27553 4650.47391 429.03353 2281.42857         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN
4 145082.51684 5201.38772 749.86044 2781.69752         NaN         NaN         NaN         NaN         NaN         NaN          NaN          NaN          NaN          NaN

In [14]:
df_control.shape

(40, 14)

In [15]:
df_test.shape

(40, 14)

In [18]:
# Let's look at the dataset

In [16]:
df_control.describe().T

count         mean         std         min         25%         50%          75%          max
Impression  40.00000 101711.44907 20302.15786 45475.94296 85726.69035 99790.70108 115212.81654 147539.33633
Click       40.00000   5100.65737  1329.98550  2189.75316  4124.30413  5001.22060   5923.80360   7959.12507
Purchase    40.00000    550.89406   134.10820   267.02894   470.09553   531.20631    637.95709    801.79502
Earning     40.00000   1908.56830   302.91778  1253.98952  1685.84720  1975.16052   2119.80278   2497.29522
Unnamed: 4   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 5   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 6   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 7   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 8   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 9   0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 10  0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 11  0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 12  0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN
Unnamed: 13  0.00000          NaN         NaN         NaN         NaN         NaN          NaN          NaN

In [17]:
df_test.describe().T

count         mean         std         min          25%          50%          75%          max
Impression  40.00000 120512.41176 18807.44871 79033.83492 112691.97077 119291.30077 132050.57893 158605.92048
Click       40.00000   3967.54976   923.09507  1836.62986   3376.81902   3931.35980   4660.49791   6019.69508
Purchase    40.00000    582.10610   161.15251   311.62952    444.62683    551.35573    699.86236    889.91046
Earning     40.00000   2514.89073   282.73085  1939.61124   2280.53743   2544.66611   2761.54540   3171.48971
Unnamed: 4   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 5   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 6   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 7   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 8   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 9   0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 10  0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 11  0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 12  0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN
Unnamed: 13  0.00000          NaN         NaN         NaN          NaN          NaN          NaN          NaN

In [19]:
# Editing the names of control and test group variables

In [20]:
df_control.columns = [i + "_control" for i in df_control.columns]
df_control.head()

Impression_control  Click_control  Purchase_control  Earning_control  Unnamed: 4_control  Unnamed: 5_control  Unnamed: 6_control  Unnamed: 7_control  Unnamed: 8_control  Unnamed: 9_control  Unnamed: 10_control  Unnamed: 11_control  Unnamed: 12_control  Unnamed: 13_control
0         82529.45927     6090.07732         665.21125       2311.27714                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN
1         98050.45193     3382.86179         315.08489       1742.80686                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN
2         82696.02355     4167.96575         458.08374       1797.82745                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN
3        109914.40040     4910.88224         487.09077       1696.22918                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN
4        108457.76263     5987.65581         441.03405       1543.72018                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN

In [21]:
df_test.columns = [i + "_test" for i in df_test.columns]
df_test.head()

Impression_test  Click_test  Purchase_test  Earning_test  Unnamed: 4_test  Unnamed: 5_test  Unnamed: 6_test  Unnamed: 7_test  Unnamed: 8_test  Unnamed: 9_test  Unnamed: 10_test  Unnamed: 11_test  Unnamed: 12_test  Unnamed: 13_test
0     120103.50380  3216.54796      702.16035    1939.61124              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
1     134775.94336  3635.08242      834.05429    2929.40582              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
2     107806.62079  3057.14356      422.93426    2526.24488              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
3     116445.27553  4650.47391      429.03353    2281.42857              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
4     145082.51684  5201.38772      749.86044    2781.69752              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN

In [29]:
# Step 2: Analyze control and test group data.
# In the Impression breakdown, by taking the sums of the clicks and the average of the purchases into account, a table has emerged according to the impressions.

In [26]:
df_control.groupby("Impression_control").agg({"Click_control": "sum",
                              "Purchase_control": "mean"})

Click_control  Purchase_control
Impression_control                                 
45475.94296            7370.21438         400.18883
77773.63390            4462.20659         519.66966
79498.24866            6653.84552         470.50137
80254.33164            3075.31120         530.67988
81872.98572            7959.12507         729.16155
82067.89994            6609.18343         267.02894
82529.45927            6090.07732         665.21125
82696.02355            4167.96575         458.08374
83676.60243            7153.97419         487.82877
83676.93601            4273.40021         386.09788
86409.94180            4608.25621         345.04603
87939.42225            3993.31926         613.71231
92044.50011            4667.20523         729.36526
94138.62750            6994.18410         593.02579
94225.52052            5255.63866         619.86719
94472.19659            5937.47947         686.41229
94923.68370            5678.98196         468.87803
95110.58627            3555.58067         512.92875
96122.65762            7453.15082         513.57883
98050.45193            3382.86179         315.08489
101530.95023           5626.90123         555.00202
101616.46555           5323.54825         801.79502
101997.49410           4736.35337         474.61354
105493.02321           2189.75316         666.48207
106649.18307           4358.02704         747.02012
108051.82405           5919.24497         521.25811
108457.76263           5987.65581         441.03405
109914.40040           4910.88224         487.09077
110479.80682           3585.03168         531.73273
114341.40155           3806.14580         620.13588
117827.06153           5560.36766         739.53719
121085.88122           4285.17861         590.40602
122709.71659           5091.55896         745.98568
123678.93423           3649.07379         476.16813
123961.86872           4898.78841         585.63917
129146.68958           5537.60924         458.15592
132064.21900           3747.15754         551.07241
132781.01987           5830.00652         628.87237
136171.27165           5784.32193         772.03288
147539.33633           3856.72685         329.37703

In [28]:
df_test.groupby("Impression_test").agg({"Click_test": "sum",
                              "Purchase_test": "mean"})

Click_test  Purchase_test
Impression_test                           
79033.83492      4495.42818      425.35910
79234.91193      6002.21358      382.04712
83356.58756      6019.69508      570.43213
96331.36543      3860.62836      889.91046
97507.36685      4119.21862      670.52139
102257.45409     4800.06832      521.31073
104971.22276     3563.51505      699.28728
106116.43664     3279.47297      491.61453
107806.62079     3057.14356      422.93426
109569.87771     2268.94706      346.46262
113732.66846     3251.84904      610.74232
115303.70814     4096.22493      453.91674
115923.00695     4213.86862      778.37316
115934.85439     5059.85845      653.49152
116445.27553     4650.47391      429.03353
116481.87337     4702.78247      472.45373
117281.76759     2617.80386      372.12579
117442.86465     3907.93924      660.47791
117989.24638     3411.71149      854.12943
118704.64150     2937.36032      531.97083
119877.96005     3622.93635      689.15574
120103.50380     3216.54796      702.16035
120513.45712     3409.26771      604.92335
122860.39591     3652.20842      357.23397
123798.62140     3938.12794      523.79992
124669.00340     1836.62986      822.59992
125957.11610     4690.56991      855.71980
129801.48637     4243.79074      628.92042
130702.23941     3626.32007      449.82459
131271.71560     4721.18781      532.27934
134387.16892     4986.10073      417.99034
134775.94336     3635.08242      834.05429
137222.38146     4042.27828      677.27270
137231.17523     3991.67745      311.62952
140219.55203     5232.71182      524.90806
141367.84841     3924.59166      501.00867
145082.51684     5201.38772      749.86044
152941.71903     3209.69344      414.22367
157681.15975     4468.26679      701.58760
158605.92048     2736.41023      748.49661

In [30]:
# Step 3: After the analysis process, combine the control and test group data using the concat method.
# In this section, control and test data combined and It's becoming much more easier to analyze.

In [33]:
df_last = pd.concat([df_control, df_test], axis=1)
df_last.head()

Impression_control  Click_control  Purchase_control  Earning_control  Unnamed: 4_control  Unnamed: 5_control  Unnamed: 6_control  Unnamed: 7_control  Unnamed: 8_control  Unnamed: 9_control  Unnamed: 10_control  Unnamed: 11_control  Unnamed: 12_control  Unnamed: 13_control  Impression_test  Click_test  Purchase_test  Earning_test  Unnamed: 4_test  Unnamed: 5_test  Unnamed: 6_test  Unnamed: 7_test  Unnamed: 8_test  Unnamed: 9_test  Unnamed: 10_test  Unnamed: 11_test  Unnamed: 12_test  Unnamed: 13_test
0         82529.45927     6090.07732         665.21125       2311.27714                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN     120103.50380  3216.54796      702.16035    1939.61124              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
1         98050.45193     3382.86179         315.08489       1742.80686                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN     134775.94336  3635.08242      834.05429    2929.40582              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
2         82696.02355     4167.96575         458.08374       1797.82745                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN     107806.62079  3057.14356      422.93426    2526.24488              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
3        109914.40040     4910.88224         487.09077       1696.22918                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN     116445.27553  4650.47391      429.03353    2281.42857              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN
4        108457.76263     5987.65581         441.03405       1543.72018                 NaN                 NaN                 NaN                 NaN                 NaN                 NaN                  NaN                  NaN                  NaN                  NaN     145082.51684  5201.38772      749.86044    2781.69752              NaN              NaN              NaN              NaN              NaN              NaN               NaN               NaN               NaN               NaN

In [42]:
# For analyzing Confidence Intervals, using sms.DescrStatsW will be used.
print(sms.DescrStatsW(df_control["Purchase_control"]).tconfint_mean())
print(sms.DescrStatsW(df_test["Purchase_test"]).tconfint_mean())

(508.0041754264924, 593.7839421139709)
(530.5670226990063, 633.645170597929)


In [34]:
# Task 2: Defining the Hypothesis of A/B Testing

In [35]:
# Adım 1: Defining the hypothesis.
# H0 : M1 = M2
# H1 : M1!= M2 

In [43]:
# Assumption of Normality
# Variance Homogeneity

# H0: There is no statistically significant difference between Average bidding and Maximum bidding in Purchase averages.
# H1 There is a statistically significant difference between Average bidding and Maximum bidding in Purchase averages

In [44]:
# Step 2: Analyze the purchase averages for the control and test group.

In [46]:
df_last[["Purchase_control", "Purchase_test"]].mean()

Purchase_control   550.89406
Purchase_test      582.10610
dtype: float64

In [50]:
# When outcomes examines, It's seen that the new promotion type which is "average bidding" haf an affect. Otherwise, we cannot say it's enough to present it as an outcome. Surely we should keep going through our hypothesis.

In [48]:
# Task 3: Performing Hypothesis Testing

Step 1: Perform hypothesis checks before hypothesis testing.
These are Assumption of Normality and Homogeneity of Variance. Test separately whether the control and test groups comply with the normality assumption, over the Purchase variable.
Normality Assumption:
H0: Normal distribution assumption is provided.
H1: The assumption of normal distribution is not provided.
p < 0.05 H0 REJECT , p > 0.05 H0 CANNOT REJECT
Is the assumption of normality according to the test result provided for the control and test groups? Interpret the p-values obtained.
Variance Homogeneity:
H0: Variances are homogeneous.
H1: Variances are not homogeneous.
p < 0.05 H0 REJECT , p > 0.05 H0 CANNOT REJECT
Test whether the homogeneity of variance is provided for the control and test groups over the Purchase variable.
Is the assumption of normality provided according to the test result? Interpret the p-values obtained.

In [56]:
#Normality Assumption (Shapiro-Wilks Test)

# For Control
test_stat, pvalue = shapiro(df_last["Purchase_control"])
print('Test Stat = %.4f, p-value = %.4f' % (test_stat, pvalue))

Test Stat = 0.9773, p-value = 0.5891


In [57]:
# For Test
test_stat, pvalue = shapiro(df_last["Purchase_test"])
print('Test Stat = %.4f, p-value = %.4f' % (test_stat, pvalue))

Test Stat = 0.9589, p-value = 0.1541


If we look at the results of the Assumption of Normality, the p-values for both control and test groups are greater than 0.05. In this case, the result is that the H0 hypothesis cannot be rejected because the p-value is greater than 0.05. Since H0 could not be rejected, the assumption of normality was provided.

In [58]:
 # Step 2: Select the appropriate test according to the Normality Assumption and Variance Homogeneity results.

In [59]:
# Variance Homogeneity (Levene Test)
# H0: Variances are homogeneous.
# H1: Variances are not homogeneous.

In [60]:
test_stat, pvalue = levene(df_last["Purchase_control"],
                           df_last["Purchase_test"])
print("Test stat = %.4f, p-value = %.4f" % (test_stat, pvalue))

Test stat = 2.6393, p-value = 0.1083


Two samples t (Parametric test) The p-value value obtained as a result of Levene's test is greater than 0.05 as observed. The result in this case is that H0 is undeniable. The H0 hypothesis reveals that the variances are homogeneous. According to the result, we could not reject H0 and variance homogeneity was achieved.

In [61]:
# Step 3: Considering the p_value obtained as a result of the test, interpret whether there is a statistically significant difference between the purchasing averages of the control and test groups.

Since we cannot reject the assumption hypotheses, the process will continue with the Parametric test.

In [62]:
test_stat, pvalue = ttest_ind(df_last["Purchase_control"],
                              df_last["Purchase_test"],
                              equal_var=True)
print("Test stat = %.4f, p-value = %.4f" % (test_stat, pvalue))
# The reason of "equal_var = True" is the variance homogeneity.

Test stat = -0.9416, p-value = 0.3493


The p-value resulting from the parametric test is 0.3493. If the P-value is higher than 0.05, it shows that the H0 hypothesis is undeniable. When the values were analyzed within the scope of the project, it was seen that the homogeneity of the variances and the H0 hypotheses were undeniable. The fact that the H0 hypothesis cannot be rejected in the parametric test, which is the last test conducted in this context, proves the statement that "There is no statistically significant difference between the Purchase averages of average bidding and Max. bidding". In the first analysis made within the scope of this project, we said that there is a difference between the newly developed model and the previous model, but that we should examine whether the result is statistically by chance. This result shows that there is no statistically significant difference between Purchase averages of Average bidding and Maximum bidding.